### ISSUE TITLE
HTTP/2 server lacks a timeout.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8480]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8480|
|Type|enhancement|
|Created|2016-06-16 12:22:15Z|
|Branch|https://github.com/twisted/twisted/tree/8480-lukasa-h2logging|

t.w.h.HTTPChannel currently times out the connection, but H2Connection does not. It probably should! We should also consider timing out individual streams as well as the connection as a whole.

<details><summary>Searchable metadata</summary>

```
trac-id__8480 8480
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8480_lukasa_h2logging 8480-lukasa-h2logging
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1466079735844824 1466079735844824
changetime__1468489609408938 1468489609408938
version__None None
owner__Lukasa Lukasa
cc__hawkowl
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

See also [#3746](https://github.com/twisted/twisted/issues/3746).

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

Let's talk about some wrinkles.

Firstly, streams. It's possible to have a connection that is very active but has one or more streams that do not progress. That's not a good thing, so we should probably aim to time streams out if they're inactive for too long. This is *a bit* frustrating because it potentially limits the utility of using a H2 stream as a long-lived communications channel, but let's worry about that in the future. The response to timing this out is probably to RST_STREAM(REFUSED_STREAM). (REFUSED_STREAM is only appropriate in the current model where we don't actually process the stream until the entire body is received, so we may need to change that later).

Secondly, connection timeout. Here we can and should overload the standard behaviour to emit a GOAWAY before we tear the connection down. Nothing wrong with being polite. However, the error code we send should be made clear based on whether we have any open streams or not, so we'll need to make sure to check with the hyper-h2 connection object to see whether this is NO_ERROR or PROTOCOL_ERROR.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

Ok, I've provided an initial patch in pull request 196: https://github.com/twisted/twisted/pull/196

This patch currently just deals with the simple case: timing out whole connections. It does not address the following concerns:

1. Leaving streams open. Twisted could be forced to consume a large amount of memory by having a client open 99 streams but not progress them, and then repeatedly progress one stream. This timeout would never give up on those 99 streams. We may need a per-stream timeout if that's a thing we want to deal with.
2. PING frames. This patch allows a client to keep a connection open indefinitely using PING frames: there's no threshold beyond which the connection will be torn down. That potentially allows a DDoS vector: multiple clients can open 100 streams, not progress them, and then repeatedly send PING frames to keep the connection alive and to force Twisted to maintain all that state. Part of this problem would be resolved by dealing with (1), but we should consider whether after a certain amount of time with nothing but PING frames we should tear the connection down anyway.

Regardless, the first part of this patch is available for review. I'd like feedback on the other parts too.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|


The branch looks very good.

Since you have write access, please create a branch in Twisted and send it to buildbot.

------

I think that is excellent to start with a timeout for the whole connection and to also update the timeout on PING or any other message... baby steps :)

-------

From the https://http2.github.io/http2-spec/#PING documentation it is not clear whether a PING frame should reset the timeout counter... but I think that this is specific to each implementation and later we can have a patch which will not reset the timeout on PING requests.

I make legitimate usage of NOOP messages for my implementation to have a high latency on new messages and to also check if the connectivity with a remote peer is still active.

-----

Please use the new logger 

--------

... getting late here so in case it is not already reviewed I will continue later

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

In changeset 2c137008d7e3df8355a1c81925a1e901e1dd14ee
```
#!CommitTicketReference repository="" revision="2c137008d7e3df8355a1c81925a1e901e1dd14ee"
Add topfile for #8480.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

Ok, I've run the builds: all green. Also confirmed coverage: all changes covered. And I've updated to use the new logger.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Hi,

Thanks for the changes.

They look good.

Only minor comment.

1. Instead of log.info() I think that the pattern is to use self._log.info . See [#8036](https://github.com/twisted/twisted/issues/8036)

2. Instead of generic # TODO: comments, please create a ticket and use `# FIXME: TICKET_URL` as a marker... in the ticket you can add as much information as you want in any style you want.

3. In `timeoutConnection` I don't understand the `# In our override of timeoutConnection` comment which precedes the log.info call. Maybe this should be part of the docstring.

4. For `# This monkeypatch + conn.callLater = reactor.callLater` maybe we can refactor it into `patch_TimeoutMixin_clock(conn, reactor)` ... and have the comment as docstring of that method.

5. For '# one byte at a time + preamble = transport.value` I would refactor it into a helper method like get_sent_data(transport, conn)` ... as the way in which that data is received should not be important for the timeout test. The whole preamble setup looks complicated and not very relevant to the test... so I would prefer to refactor that code as helpers.

6. `# We disconnected after sending a GoAway frame.` maybe `# Connection is disconnected due to timeout and  a final NO_ERROR message is sent before disconnecting`

7. Do we really need the log ? If yes, maybe we should also have a test that the logging is generated , but maybe this should be discussed in [#8036](https://github.com/twisted/twisted/issues/8036)


Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|GitHub set status to `closed`|
|-|-|

In changeset 08b55a87f42250755ff24ab1a2380969f7cc427b
```
#!CommitTicketReference repository="" revision="08b55a87f42250755ff24ab1a2380969f7cc427b"
Merge 8480-lukasa-h2logging: Time out HTTP/2 connections.

Author: lukasa
Reviewer: adiroiban
Fixes: #8480
```
